mobile_api_guidance.md

```diff
@@ -0,0 +1,7 @@
+# Versioning
+
+Functionality provided by an API inevitably changes over time, and it’s sometimes necessary to make breaking, backwards-incompatible changes. In contrast to web app users, mobile app users must upgrade their software to take advantage of new functionality, and there is a period of overlap and adoption where two versions of a client are in circulation. It is beneficial to think about upgrade strategies early in your API’s development, and there are a few different approaches.
+
+The most common approach is to prefix URIs with a version, e.g. ‘/v1/method’. This allows clients that are configured to use a particular version of an API to continue using it after a new version of the API is released with a different prefix. Facebook is an example of an API that is versioned this way.
+
```

Comment:
@intrepidmatt woops, missed this comment.

We follow Heroku's API design guidelines, which say to version in headers: https://github.com/interagent/http-api-design#require-versioning-in-the-accepts-header

There is a lot of noise on the web generally about URL versioning being bad, although I've found little in the way of persuasive arguments about this and don't have super strong feelings either way.  The best argument I've seen is perhaps this one, from this blog post: https://www.mnot.net/blog/2012/12/04/api-evolution


> One viewpoint is to say that the service is being versioned, the service is identified by the URL, and therefore the > version goes in the URL, like this:

> http://api.example.com/v1.2.3/ladder

> However, there are (at least) two issues to consider with this approach.

> First of all, it's coarse-grained, in that you can't evolve parts of the system independently. For example, 
> introducing a new format for the "ladder" resource - by the rules, that means a minor version, which means that 
> the URL should now become:

> http://api.example.com/v1.3.0/ladder

> which, as discussed previously in the API Versioning Smackdown, creates a whole new tree of resources, and 
> tightly couples the client and server.

> While that may be fine if you have a very simple API with no interdependencies -- such as serving a bunch of 
> JavaScript, as Nicholas Zakas describes -- it will quickly become a huge headache for testing, support and 
> operations if you have to support all of the combinations of possible resources and their interactions in a more 
> complex one.

> Second, it's also intermingling the version into identifiers. Because URIs are used in the Web as the fundamental 
> identifier by so many things -- caches, spiders, forms, and so on -- embedding information that's likely to change 
> into them make them unstable, thereby reducing their value. For example, a cache invalidates content associated 
> with a URL when a POST request flows through it; if the URL is different because there are different versions 
> floating around, it doesn't work.